src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py

```diff
@@ -63,9 +63,60 @@ def float_nan_inf_to_sql(value: float) -> str:
     return f"{cast_value} :: FLOAT"
 
 
-def to_sql(value: Any, datatype: DataType, from_values_statement: bool = False) -> str:
+def to_sql(
+    value: Any,
+    datatype: DataType,
+    from_values_statement: bool = False,
+    is_system_function: bool = False,
+) -> str:
     """Convert a value with DataType to a snowflake compatible sql"""
-
+    # if is system function, don't do covert
+    if is_system_function:
```

Comment:
I guess if we are creating an else branch which does this, instead we should create a new function `to_sql_no_cast` and call this when calling from any function that may be system function or any other that should skip casting.